scanpointgenerator/core/compoundgenerator.py

```diff
@@ -21,236 +23,315 @@ def __init__(self, generators, excluders, mutators):
-        self.periods = []
-        for generator in self.generators[::-1]:
-            self.num *= generator.num
-            self.periods.insert(0, self.num)
-
-        logging.debug("CompoundGenerator periods")
-        logging.debug(self.periods)
-
-        if self.excluders:  # Calculate number of remaining points and flatten
-                            # index dimensions
-            remaining_points = 0
-            for _ in self._filtered_base_iterator():
-                # TODO: Faster with enumerate()?
-                remaining_points += 1
-            self.index_dims = [remaining_points]
-            self.num = remaining_points
-
         if len(self.axes) != len(set(self.axes)):
-            raise ValueError("Axis names cannot be duplicated; given %s" %
-                             self.index_names)
-
-        # These are set when using the get_point() interface
-        self._cached_iterator = None
-        self._cached_points = []
-        self._cached_lock = Lock()
-
-    def _get_sub_point(self, gen_index, point_num):
-        points = self.point_sets[gen_index]
-        axis_period = self.periods[gen_index]
-        axis_length = len(points)
-        # Can't use index_dims in case they have been flattened
-        # by an excluder
-
-        point_index = \
-            (point_num / (axis_period / axis_length)) % axis_length
-        loop_number = point_num / axis_period
+            raise ValueError("Axis names cannot be duplicated")
 
-        # Floor floats to ints for indexing
-        point_index = int(point_index)
-        loop_number = int(loop_number)
-        if self.alternate_direction[gen_index] and loop_number % 2:
-            point_index = (axis_length - 1) - point_index
-            reverse = True
-        else:
-            reverse = False
-
-        sub_point = points[point_index]
-
-        return reverse, sub_point
-
-    def _base_iterator(self):
-        """
-        Iterator to generate points by nesting each generator in self.generators
-
-        Yields:
-            Point: Base points
-        """
-        num_point_sets = len(self.point_sets)
-        for point_num in range_(self.num):
-            point = Point()
-            for gen_index in range_(num_point_sets - 1):
-                reverse, sub_point = self._get_sub_point(gen_index, point_num)
+        self.generators = generators
+        self.generator_dim_scaling = {}
 
-                # Outer indexes use positions
-                point.positions.update(sub_point.positions)
-                point.upper.update(sub_point.positions)
-                point.lower.update(sub_point.positions)
-                point.indexes += sub_point.indexes
+    def prepare(self):
+        self.num = 1
+        self.dimensions = []
+        # we're going to mutate these structures
+        excluders = list(self.excluders)
+        generators = list(self.generators)
+
+        # special case if we have rectangular regions on line generators
```

Comment:
Apparently the users specify in the GUI a bounding box (grid, circle, polygon), and a fill pattern (raster, spiral, lissajous), and there are multiple bits of GUI code that can do this, so it's better to detect the grid in a rectangle bit here...